Backend/Remora.Discord.API/Json/Converters/Internal/PayloadConverter.cs

```diff
@@ -252,13 +253,44 @@ private RetrieveEntityResult<OperationCode> GetOperationCode(Type objectType)
             };
         }
 
-        private IPayload DeserializePayload<TData>(JsonElement dataProperty, JsonSerializerOptions options)
+        private static IPayload DeserializePayload<TData>(JsonElement dataProperty, JsonSerializerOptions options)
             where TData : IGatewayPayloadData
         {
             var data = JsonSerializer.Deserialize<TData>(dataProperty.GetRawText(), options);
+
+            if (data is null)
+            {
+                throw new JsonException();
+            }
+
             return new Payload<TData>(data);
         }
 
+        private static IPayload DeserializeHeartbeatPayload(JsonElement dataProperty, JsonSerializerOptions options)
+        {
+            var data = JsonSerializer.Deserialize<long?>(dataProperty.GetRawText(), options);
+
+            return new Payload<IHeartbeat>(new Heartbeat(data));
+        }
+
+        private static IPayload DeserializeInvalidSessionPayload(JsonElement dataProperty, JsonSerializerOptions options)
```

Comment:
This method should not be neccesary; see https://github.com/Nihlus/Remora.Discord/blob/master/Backend/Remora.Discord.API/Json/Converters/Internal/InvalidSessionConverter.cs

Reply:
Indeed.